spark/src/main/scala/org/apache/spark/sql/delta/DeltaHistoryManager.scala

```diff
@@ -118,28 +142,92 @@ class DeltaHistoryManager(
+    val ictEnablementCommit =
+      if (DeltaConfigs.IN_COMMIT_TIMESTAMPS_ENABLED.fromMetaData(snapshot.metadata)) {
+        InCommitTimestampUtils.getValidatedICTEnablementInfo(snapshot.metadata)
+          // If missing, ICT is enabled for all available versions
+          .getOrElse(placeholderEarliestCommit)
+      } else {
+        // Pretend ICT will be enabled after the latest version and requested timestamp.
+        // This will force us to use the non-ICT search path below.
+        Commit(latestVersion + 1, time + 1)
+      }
 
-    // Search for the commit
-    val commit = if (latestVersion - earliest > 2 * maxKeysPerList) {
-      parallelSearch(time, earliest, latestVersion + 1)
+    var commitOpt = if (ictEnablementCommit.timestamp <= time) {
+      // ICT was enabled as-of the requested time
+      if (snapshot.timestamp <= time) {
+        // We just proved we should use the latest snapshot
+        Some(Commit(snapshot.version, snapshot.timestamp))
+      } else {
+        // start ICT search over [earliest available ICT version, latestVersion)
+        val searchWindowLowerBoundCommit = if (earliestVersion <= ictEnablementCommit.version) {
+          ictEnablementCommit
+        } else {
+          placeholderEarliestCommit
+        }
+        // Note that this search can return `placeholderEarliestCommit`.
+        // The real timestamp of the earliest commit will be fetched later.
+        getActiveCommitAtTimeFromICTRange(
+          time,
+          searchWindowLowerBoundCommit,
+          latestVersion + 1,
+          deltaLog.newDeltaHadoopConf(),
+          deltaLog.logPath,
+          deltaLog.store,
+          numChunks = 10,
+          spark,
+          commitFileProvider)
+      }
     } else {
-      val commits = getCommits(
-        deltaLog.store,
+      // ICT was NOT enabled as-of the requested time
+      if (ictEnablementCommit.version <= earliestVersion) {
+        // We're searching for a non-ICT time but the non-ICT commits are all missing.
+        // If `canReturnEarliestCommit` is `false`, we need the details of the
+        // earliest commit to populate the TimestampEarlierThanCommitRetentionException
+        // error correctly.
+        // Else, when `canReturnEarliestCommit` is `true`, the earliest commit
+        // is the desired result.
+        // The real timestamp of the earliest commit will be fetched later.
+        Some(placeholderEarliestCommit)
+      } else {
+        // start non-ICT search over [earliestVersion, ictEnablementVersion)
+        Some(getCommitFromNonICTRange(earliestVersion, end = ictEnablementCommit.version, time))
+      }
+    }
+
+    // We need to fetch the correct timestamp for the earliest commit if it was the result of the
+    // search.
+    // If commitOpt == ictEnablementCommit, we also need to validate the existence of the ICT
+    // enablement commit.
+    if (commitOpt.contains(placeholderEarliestCommit) || commitOpt.contains(ictEnablementCommit)) {
```

Comment:
Yes, the snapshot construction will fail at some step downstream.

Reply:
Is it important to verify here that the enablement commit exists, then?

Reply:
It is only *required* when `canReturnEarliest` is `true`, so that we return some valid commit in that case. Otherwise, we can probably get away with not doing this at the cost of throwing a somewhat confusing error message instead of "proper" one that exists here: https://github.com/delta-io/delta/pull/2813/files/4a08b6e6b72a1c617bee0ca046849ecd790b7075#diff-98df89d4b0ce76abb5263d4fbb6f991660083a254dcdb9d922a4973a9937a4e2L149 . Since most clients are not trying to query for the earliest version of the table, I believe that this additional IO might be a reasonable penalty to pay.